File: step17example/src/main/java/com/example/step17example/MainActivity.java
package com.example.step17example;

import android.content.DialogInterface;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import com.example.step17example.databinding.ActivityMainBinding;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
/*
    view binding 사용법
    1. build.gradle 파일에 아래의 설정 추가
        buildFeatures {
            viewBinding = true
        }
    2. gradle sync (우상단 sync now)로 설정 적용
    3. Layout xml 문서에 이름대로 클래스가 만들어진다.
        activity_main.xml  ->  ActivityMainBinding
        activity_sub.xml   ->  ActivitySubBinding
 */
public class MainActivity extends AppCompatActivity implements Util.RequestListener, AdapterView.OnItemLongClickListener {

    List<TodoDto> todo = new ArrayList<>();
    TodoAdapter adapter;
    ActivityMainBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        //setContentView(R.layout.activity_main);
        //R.layout.activity_main.xml 문서를 전개해서 View를 만들기
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        //전개된 layout에서 root를 얻어내서 화면을 구성한다. (여기서는 Linearlayout)
        setContentView(binding.getRoot());

        adapter = new TodoAdapter(this, R.layout.listview_cell, todo);
        binding.listView.setAdapter(adapter);

        binding.addBtn.setOnClickListener(v -> {
            String content = binding.inputText.getText().toString();
            Map<String, String> map = new HashMap<>();
            map.put("content", content);
            Util.sendPostRequest(
                    AppConstants.REQUEST_TODO_INSERT,
                    AppConstants.BASE_URL+"/todo/insert",
                    map,
                    this
            );
            binding.inputText.setText("");
        });

        binding.listView.setOnItemLongClickListener(this);
    }

    @Override
    protected void onStart() {
        super.onStart();
        //MainActivity가 활성화되는 시점에 List 불러오기
        Util.sendGetRequest(
                AppConstants.REQUEST_TODO_LIST,
                AppConstants.BASE_URL+"/todo/list",
                null,
                this
        );
    }

    @Override
    public void onSuccess(int requestId, Map<String, Object> result) {
        String jsonStr = (String)result.get("data");

        if (requestId == AppConstants.REQUEST_TODO_INSERT){
            Log.d("MainActivity onSuccess()", jsonStr);
            Util.sendGetRequest(
                    AppConstants.REQUEST_TODO_LIST,
                    AppConstants.BASE_URL+"/todo/list",
                    null,
                    this
            );
            binding.listView.smoothScrollToPosition(adapter.getCount());
        }else if(requestId == AppConstants.REQUEST_TODO_LIST){
            todo.clear();
            try {
                JSONArray arr = new JSONArray(jsonStr);
                for(int i = 0; i < arr.length(); i++){
                    JSONObject tmp = arr.getJSONObject(i);

                    int num = tmp.getInt("num");
                    String content = tmp.getString("content");
                    String regdate = tmp.getString("regdate");
                    todo.add(new TodoDto(num, content, regdate));
                }
                adapter.notifyDataSetChanged();
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        }else if(requestId == AppConstants.REQUEST_TODO_DELETE){
            Util.sendGetRequest(
                    AppConstants.REQUEST_TODO_LIST,
                    AppConstants.BASE_URL+"/todo/list",
                    null,
                    this
            );
        }
    }

    @Override
    public void onFail(int requestId, Map<String, Object> result) {
        if(requestId == AppConstants.REQUEST_TODO_INSERT){

        }else if(requestId == AppConstants.REQUEST_TODO_LIST){

        }else if(requestId == AppConstants.REQUEST_TODO_DELETE){

        }
    }

    @Override
    public boolean onItemLongClick(AdapterView<?> parent, View view, int position, long id) {
        String num = Long.toString(id);

        new AlertDialog.Builder(this)
                .setTitle("알림")
                .setMessage("삭제하시겠습니까?")
                .setPositiveButton("네", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        Map<String, String> map = new HashMap<>();
                        map.put("num", num);
                        Util.sendPostRequest(
                                AppConstants.REQUEST_TODO_DELETE,
                                AppConstants.BASE_URL+"/todo/delete",
                                map,
                                MainActivity.this
                        );
                    }
                })
                .setNegativeButton("아니오", null)
                .create()
                .show();
        return false;
    }
}

File: step17httprequest3/src/main/java/com/example/step17httprequest3/MainActivity.kt
package com.example.step17httprequest3

import android.os.Bundle
import android.util.Log
import android.widget.Button
import android.widget.EditText
import androidx.appcompat.app.AppCompatActivity
import org.json.JSONArray
import org.json.JSONObject

class MainActivity : AppCompatActivity() , Util.RequestListener{
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        //EditText 객체의 참조값 얻어오기
        val editText = findViewById<EditText>(R.id.inputMsg)
        //Button 객체의 참조값 얻어오기
        val sendBtn = findViewById<Button>(R.id.sendBtn)
        sendBtn.setOnClickListener {
            //EditText에 입력한 문자열을 읽어와서
            val msg = editText.text.toString()
            Util.sendPostRequest(
                    999,
                    "http://192.168.0.29:9000/boot07/api/send",
                    mapOf("msg" to msg), //"msg" 라는 키 값으로 입력한 메세지를 담은 Map
                    this
            )
        }

        val getListBtn = findViewById<Button>(R.id.getListBtn)
        getListBtn.setOnClickListener {
            Util.sendGetRequest(
                    1000,
                    "http://192.168.0.29:9000/boot07/api/list",
                    mapOf("pageNum" to "1"),
                    this
            )
        }
    }

    override fun onSuccess(requestId: Int, result: Map<String, Any?>?) {
        if(requestId == 999){
            //웹서버가 응답한 json 문자열 {}
            var jsonStr = result?.get("data").toString()
            Log.d("#### json 문자열 ####", jsonStr)
            // JSONObject 객체를 생성하면서 생성자의 인자로 json 문자열을 전달
            val obj = JSONObject(jsonStr)
            val isSuccess = obj.getBoolean("isSuccess")
            val response = obj.getString("response")
            val num = obj.getInt("num")
        }else if(requestId == 1000){
            val jsonStr = result?.get("data").toString()
            Log.d("#### json 문자열 ####", jsonStr)
            val arr = JSONArray(jsonStr)
            //i 값을 0에서 부터 JSONArray의 사이즈 -1 까지 변화시키며 반복
            for(i in 0 .. arr.length() - 1){
                val tmp = arr.getString(i)
                Log.d("json array", tmp)
            }
        }
    }

    override fun onFail(requestId: Int, result: Map<String, Any?>?) {

    }
}